### ISSUE TITLE
Version 2 of string interning

### ISSUE THREAD

**Issue by GiedriusS** [thanos-io/thanos#7067]
### Is your proposal related to a problem?

We currently have interning in Receive under a feature flag. It works but it could be improved a lot. We still allocate memory for each duplicate and then deduplicate later on.

Remote write requests and series responses will allocate new memory for each duplicated string. The most obvious pathological case is when querying histograms. The label `le` can be easily allocated thousands of times. This is even worse because of incoming duplicated data due to HA setup.

I think this leads to huge inefficiencies.

### Describe the solution you'd like

We should avoid allocating memory for duplicated strings. This could be a per-request map of hash to []byte/string.

We should be able to use this same deduplication technique in remote write, series responses, and any other place.

I think gRPC doesn't allow accessing the byte streams as it is a message oriented format so I think this entails using something else.

**Comment by bwplotka** [comment]
BTW this problem will be totally fixed by Remote Write 2.0 https://docs.google.com/document/d/1PljkX3YLLT-4f7MqrLt7XCVPG3IsjRREzYrUzBxCPV0/edit#heading=h.3p42p5s8n0ui

**Comment by GiedriusS** [comment]
That's true but I'd like to solve this problem at the RPC level once and for all so that all and future hot RPC paths would be protected from this issue automatically. Otherwise, we will have to implement this https://github.com/thanos-io/thanos/pull/5906, have a list of string refs like your proposal in the remote_write calls, etc.

**Comment by ringerc** [comment]
A means to limit the hash size might make sense here. Unless the hash is always going to be smaller than the total memory consumption without the hash.

**Comment by GiedriusS** [comment]
I will start over soon with much simpler versions first e.g. having some list of common label names baked into the binary so that we could avoid sending the symbol table over the wire.

**Comment by ringerc** [comment]
Interesting idea. Won't help with label values though.

I don't mean to block this work in any way. Just so long as there's a way to limit the symbol table size, and/or turn this off for memory-constrained systems that handle high cardinality data.
